src/rebuild/srv.c

```diff
@@ -595,44 +601,42 @@ rebuild_leader_status_check(struct ds_pool *pool, uint32_t map_ver, uint32_t op,
 
 	while (1) {
 		struct daos_rebuild_status	*rs = &rgt->rgt_status;
-		struct pool_target		*targets;
 		char				sbuf[RBLD_SBUF_LEN];
-		unsigned int			failed_tgts_cnt;
 		double				now;
 		char				*str;
+		d_rank_list_t			excluded = { 0 };
+		int				i;
 
-		rc = pool_map_find_failed_tgts(pool->sp_map, &targets,
-					       &failed_tgts_cnt);
+		ABT_rwlock_rdlock(pool->sp_lock);
+		rc = map_ranks_init(pool->sp_map,
+				    PO_COMP_ST_UP | PO_COMP_ST_DOWN | PO_COMP_ST_DOWNOUT,
+				    &excluded);
 		if (rc != 0) {
-			D_ERROR("failed to create failed tgt list: "DF_RC"\n",
-				DP_RC(rc));
-			break;
+			D_INFO(DF_UUID": get rank list: %d\n", DP_UUID(pool->sp_uuid), rc);
+			ABT_rwlock_unlock(pool->sp_lock);
+			goto sleep;
 		}
 
-		if (targets != NULL) {
+		for (i = 0; i < excluded.rl_nr; i++) {
 			struct pool_domain *dom;
-			int i;
-
-			for (i = 0; i < failed_tgts_cnt; i++) {
-				dom = pool_map_find_node_by_rank(pool->sp_map,
-						targets[i].ta_comp.co_rank);
-
-				D_ASSERT(dom != NULL);
-				D_DEBUG(DB_REBUILD, "rank %d/%x.\n",
-					dom->do_comp.co_rank,
-					dom->do_comp.co_status);
-				if (pool_component_unavail(&dom->do_comp, false))
-					rebuild_leader_set_status(rgt, dom->do_comp.co_rank,
-								  -1, SCAN_DONE | PULL_DONE);
-			}
-			D_FREE(targets);
-		}
 
-		if (myrank != pool->sp_iv_ns->iv_master_rank &&
-		    pool->sp_iv_ns->iv_master_rank != -1)
-			D_DEBUG(DB_REBUILD, DF_UUID" leader is being changed"
-				" %u->%u.\n", DP_UUID(pool->sp_uuid), myrank,
-				pool->sp_iv_ns->iv_master_rank);
+			dom = pool_map_find_node_by_rank(pool->sp_map, excluded.rl_ranks[i]);
+			D_ASSERT(dom != NULL);
+
+			/* If the rank is being scheduled after reintegration
+			 * start, let's skip the rank, i.e. set it DONE.
+			 */
+		    	if (dom->do_comp.co_status == PO_COMP_ST_UP &&
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs